### PR TITLE
SI-7035 Centralize case field accessor sorting.

### PR DESCRIPTION
**Pull request by retronym**
It is both burdensome and dangerous to expect callers
to reorder these. This was seen in the field permutation
in the unapply method; a regression in 2.10.0.

Review by @paulp

### CODE REVIEW

File: src/reflect/scala/reflect/internal/Symbols.scala [hunk 1]
```diff
@@ -1728,8 +1728,27 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
...
      *  The empty list for all other classes.
-     */
-    final def caseFieldAccessors: List[Symbol] =
+     *
+     * This list will be sorted to correspond to the declaration order
+     * in the constructor parameter
+     */
+    final def caseFieldAccessors: List[Symbol] = {
+      // We can't rely on the ordering of the case field accessors within
+      // decls, handling of non-public parameters knows out the order. (See SI-7035)
```
Comment (JamesIry) [inline src/reflect/scala/reflect/internal/Symbols.scala]: Typo: "knows out the order"

Comment (retronym) [inline src/reflect/scala/reflect/internal/Symbols.scala]: Amended, thanks.

File: src/reflect/scala/reflect/internal/Symbols.scala [hunk 2]
```diff
@@ -1728,8 +1728,27 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
...
+      // (need to undo name-mangling, including the sneaky trailing whitespace)
+      //
+      // The slightly more principled approach of using the paramss of the
+      // primary constructor leads to cycles in, for example, pos/t5084.scala.
+      val primaryNames = constrParamAccessors.map(acc => nme.dropLocalSuffix(acc.name))
+      caseFieldAccessorsUnsorted.sortBy { acc =>
+        primaryNames indexWhere { orig =>
+          (acc.name == orig) || (acc.name startsWith (orig append "$"))
+        }
+      }
```
Comment (paulp) [inline src/reflect/scala/reflect/internal/Symbols.scala:1749]: This looks ripe for generalization (I haven't looked around for what other ad hoc variations of this are being performed, but I will bet my sweet bippy they're out there.) Not necessary for this PR, just throwing it out there.

**Comment by paulp** [comment]
LGTM.

File: src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala [hunk 3]
```diff
@@ -746,26 +746,8 @@ trait PatternMatching extends Transform with TypingTransformers with ast.TreeDSL
 
       // reference the (i-1)th case accessor if it exists, otherwise the (i-1)th tuple component
       override protected def tupleSel(binder: Symbol)(i: Int): Tree = { import CODE._
-        // caseFieldAccessors is messed up after typers (reversed, names mangled for non-public fields)
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala:749]: Geez @moors, take a look around trunk, I've written some code you can use if you're feeling crazy. (Didn't this logic have to exist somewhere for the old pattern matcher?)

**Pull request merged**